plutus-core/testlib/PlutusCore/Generators/PIR/Substitutions.hs

```diff
@@ -0,0 +1,211 @@
+        (TyVar _ x, b) | validSolve x b                -> pure $ Map.insert x b sub
+        (a, TyVar _ y) | validSolve y a                -> pure $ Map.insert y a sub
+        (TyFun _ a1 a2, TyFun _ b1 b2 )                -> unifies sub locals [a1, a2] [b1, b2]
+        (TyApp _ a1 a2, TyApp _ b1 b2 )                -> unifies sub locals [a1, a2] [b1, b2]
+        (TyBuiltin _ c1, TyBuiltin _ c2) | c1 == c2    -> pure sub
+        (TyForall _ x k a', TyForall _ y k' b')
+          | k == k'                                    -> go sub (Set.insert z locals)
+                                                                 (renameType x z a')
+                                                                 (renameType y z b')
+          where z = freshenTyName (locals <> Map.keysSet ctx) x
+        (TyLam _ x k a', TyLam _ y k' b')
+          | k == k'                                    -> go sub (Set.insert z locals)
+                                                                 (renameType x z a')
+                                                                 (renameType y z b')
+          where z = freshenTyName (locals <> Map.keysSet ctx) x
+        _                                              -> mzero
+      where
+        -- Check that in the current context we can solve variable z to type ty.
+        validSolve z ty = and [ Set.member z flex -- z must be a "meta variable"
+                              -- z can't be a locally bound variable
+                              , not $ Set.member z locals
+                              -- we can't solve z by changing it to a type
+                              -- where `z` is free in the type.
+                              , not $ Set.member z fvs
+                              -- The solve has to be well scoped
+                              , checkKind ctx ty (ctx Map.! z)
+                              -- We can't capture a locally bound variable
+                              , null $ Set.intersection fvs locals ]
+                              where fvs = fvTypeR sub ty
+
+    unifies sub _ [] [] = pure sub
+    unifies sub locals (a : as) (b : bs) = do
+      sub1 <- go sub locals a b
+      unifies sub1 locals as bs
+    unifies _ _ _ _ = mzero
+
+-- | Parallel substitution
+parSubstType :: Map TyName (Type TyName DefaultUni ())
+             -> Type TyName DefaultUni ()
+             -> Type TyName DefaultUni ()
+parSubstType = substType' False
+
+-- CODE REVIEW: this function is a bit strange and I don't like it. Ideas welcome for how to
+-- do this better. It basically deals with the fact that we want to be careful when substituting
+-- the datatypes that escape from a term into the type. It's yucky but it works.
+--
+-- This might not be a welcome opinion, but working with this stuff exposes some of
+-- the shortcomings of the current PIR design. It would be cleaner if a PIR program was a list
+-- of declarations and datatype declarations weren't in terms.
```

Comment:
It is a welcome opinion.

I think we did it this way, because we wanted PIR to be a superset of PLC in a way that allows us to define AST-generic functions and use them for both the ASTs. @michaelpj do you happen to remember?

Frankly I'd much prefer even for PLC to be a list of declarations.

Reply:
Resolving this, because I don't think we can do anything about it at this point.